File: src/robot_life/runtime/execution_support.py
from __future__ import annotations

from time import monotonic
from typing import Any, Callable

from robot_life.runtime.telemetry import emit_stage_trace


ExecutionRecorder = Callable[[Any], None]
BehaviorSceneResolver = Callable[[str], str]


def finalize_execution(
    *,
    execution: Any,
    telemetry: Any,
    result: Any,
    stage_name: str,
    started_at: float,
    ended_at: float | None = None,
    async_mode: bool = False,
    tick_mode: bool = False,
    record_decay_execution: ExecutionRecorder,
    cooldown_manager: Any | None,
    behavior_to_scene_type: BehaviorSceneResolver,
) -> None:
    result.execution_results.append(execution)
    record_decay_execution(execution)
    if cooldown_manager is not None and execution.status in {"finished", "degraded"}:
        scene_type = execution.scene_type or behavior_to_scene_type(execution.behavior_id)
        cooldown_manager.record_execution(scene_type, target_id=execution.target_id)
    emit_stage_trace(
        telemetry,
        execution.trace_id,
        stage_name,
        payload={
            "behavior_id": execution.behavior_id,
            "status": execution.status,
            "degraded": execution.degraded,
            "interrupted": execution.interrupted,
            "async": async_mode,
            "tick_mode": tick_mode,
            "interaction_episode_id": getattr(execution, "interaction_episode_id", None),
            "scene_epoch": getattr(execution, "scene_epoch", None),
            "decision_epoch": getattr(execution, "decision_epoch", None),
        },
        started_at=started_at,
        ended_at=ended_at if ended_at is not None else monotonic(),
    )


def enqueue_resumed_decision(*, arbitration_runtime: Any | None, telemetry: Any, resumed: Any) -> bool:
    if arbitration_runtime is None:
        return False
    enqueued = arbitration_runtime.queue.enqueue(
        resumed,
        timeout_ms=arbitration_runtime.arbitrator.queue_timeout_ms(resumed.priority),
    )
    emit_stage_trace(
        telemetry,
        resumed.trace_id,
        "resume_enqueue",
        status="queued" if enqueued is not None else "dropped",
        payload={
            "target_behavior": resumed.target_behavior,
            "priority": resumed.priority.value,
            "reason": resumed.reason,
            "interaction_episode_id": getattr(resumed, "interaction_episode_id", None),
            "scene_epoch": getattr(resumed, "scene_epoch", None),
            "decision_epoch": getattr(resumed, "decision_epoch", None),
        },
        started_at=monotonic(),
        ended_at=monotonic(),
    )
    return enqueued is not None

File: Robot_Life_CPP/src/runtime/execution_support.cpp
#include "robot_life_cpp/runtime/execution_support.hpp"

#include "robot_life_cpp/event_engine/arbitration_runtime.hpp"
#include "robot_life_cpp/runtime/telemetry.hpp"

namespace robot_life_cpp::runtime {

void finalize_execution(
    std::vector<common::ExecutionResult>* execution_results,
    const common::ExecutionResult& execution,
    TelemetrySink* telemetry,
    const std::string& stage_name,
    double started_at,
    std::optional<double> ended_at,
    bool async_mode,
    bool tick_mode) {
  if (execution_results != nullptr) {
    execution_results->push_back(execution);
  }

  emit_stage_trace(
      telemetry,
      execution.trace_id,
      stage_name,
      "ok",
      {
          {"behavior_id", execution.behavior_id},
          {"status", execution.status},
          {"degraded", execution.degraded ? "true" : "false"},
          {"interrupted", execution.interrupted ? "true" : "false"},
          {"async", async_mode ? "true" : "false"},
          {"tick_mode", tick_mode ? "true" : "false"},
      },
      started_at,
      ended_at.value_or(common::now_mono()));
}

bool enqueue_resumed_decision(
    event_engine::ArbitrationRuntime* arbitration_runtime,
    TelemetrySink* telemetry,
    const common::ArbitrationResult& resumed,
    double now_mono_s) {
  if (arbitration_runtime == nullptr) {
    return false;
  }

  auto queued = arbitration_runtime->submit(
      common::SceneCandidate{
          .scene_id = common::new_id(),
          .trace_id = resumed.trace_id,
          .scene_type = "resume_scene",
          .based_on_events = {},
          .score_hint = 1.0,
          .valid_until_monotonic = now_mono_s + 1.0,
          .target_id = std::nullopt,
          .payload = {{"resume_target_behavior", resumed.target_behavior}},
      },
      common::EventPriority::P0,
      std::nullopt,
      now_mono_s);

  const bool enqueued = !queued.has_value();
  emit_stage_trace(
      telemetry,
      resumed.trace_id,
      "resume_enqueue",
      enqueued ? "queued" : "executed",
      {
          {"target_behavior", resumed.target_behavior},
          {"priority", common::to_string(resumed.priority)},
          {"reason", resumed.reason},
      },
      now_mono_s,
      now_mono_s);
  return enqueued;
}

}  // namespace robot_life_cpp::runtime
